src/circuit/gadget/utilities/lookup_range_check.rs

```diff
@@ -0,0 +1,319 @@
+    ///
+    /// The `running_sum` advice column breaks the field element into `K`-bit
+    /// words. It is used to construct the input expression to the lookup
+    /// argument.
+    ///
+    /// The `table_idx` fixed column contains values from [0..2^K). Looking up
+    /// a value in `table_idx` constrains it to be within this range. The table
+    /// can be loaded outside this helper.
+    pub fn configure(
+        meta: &mut ConstraintSystem<F>,
+        q_lookup: Column<Fixed>,
+        running_sum: Column<Advice>,
+        table_idx: Column<Fixed>,
+        perm: Permutation,
+    ) -> Self {
+        let config = LookupRangeCheckConfig {
+            q_lookup,
+            running_sum,
+            table_idx,
+            perm,
+            _marker: PhantomData,
+        };
+
+        meta.lookup(|meta| {
+            let q_lookup = meta.query_fixed(config.q_lookup, Rotation::cur());
+            let z_cur = meta.query_advice(config.running_sum, Rotation::cur());
+            let z_next = meta.query_advice(config.running_sum, Rotation::next());
+            //    z_i = 2^{K}⋅z_{i + 1} + a_i
+            // => a_i = z_i - 2^{K}⋅z_{i + 1}
+            let word = z_cur - z_next * F::from_u64(1 << K);
+            let table = meta.query_fixed(config.table_idx, Rotation::cur());
+
+            vec![(q_lookup * word, table)]
+        });
+
+        config
+    }
+
+    #[cfg(test)]
+    // Loads the values [0..2^K) into `table_idx`. This is only used in testing
+    // for now, since the Sinsemilla chip provides a pre-loaded table in the
+    // Orchard context.
+    fn load(&self, layouter: &mut impl Layouter<F>) -> Result<(), Error> {
+        layouter.assign_region(
+            || "table_idx",
+            |mut gate| {
+                // We generate the row values lazily (we only need them during keygen).
+                for index in 0..(1 << K) {
+                    gate.assign_fixed(
+                        || "table_idx",
+                        self.table_idx,
+                        index,
+                        || Ok(F::from_u64(index as u64)),
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    // Only the lower `num_words * K` bits of the field element are constrained
+    // by this function. If the field element does not fit into this range, then
+    // the final cumulative sum `z_{num_words}` will be nonzero.
+    //
+    // It is up to the caller to constrain `z_{num_words}` == 0` outside this
+    // helper, or otherwise constrain upper bits not covered within the
+    // `num_words * K` range.
+    pub fn lookup_range_check(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        element: CellValue<F>,
+        num_words: usize,
+    ) -> Result<Vec<CellValue<F>>, Error> {
+        // `num_words` must fit into a single field element.
+        assert!(num_words <= F::NUM_BITS as usize / K);
```

Comment:
`num_words * K <= F::NUM_BITS`?

Reply:
Actually, it should be _strictly less than_ right? aka `F::CAPACITY`?

Reply:
my intention is: `K = 10`, `F::NUM_BITS = 255`, `num_words <= 255 / 10 = 25`